### ISSUE TITLE
Unable to run on heroku

### ISSUE THREAD

**Issue by niwasmala** [saleor/react-storefront#84]
I've follow the instruction on how to deploy storefront and admin. The admin is successfully deployed on heroku. But the storefront get error.

First build error I get show log like this
```
-----> Installing binaries
       engines.node (package.json):  >=14.0.0
       engines.npm (package.json):   unspecified (use default)
       
       Resolving node version >=14.0.0...
       Downloading and installing node 17.1.0...
       Using default npm version: 8.1.2
       
-----> Restoring cache
       Caching has been disabled because NODE_MODULES_CACHE=false
       
-----> Installing dependencies
       Installing node modules (package.json)
       npm ERR! code ERESOLVE
       npm ERR! ERESOLVE unable to resolve dependency tree
       npm ERR! 
       npm ERR! While resolving: react-storefront@0.1.0
       npm ERR! Found: graphql@16.0.1
       npm ERR! node_modules/graphql
       npm ERR!   graphql@"16.0.1" from the root project
       npm ERR! 
       npm ERR! Could not resolve dependency:
       npm ERR! peer graphql@"^14.0.0 || ^15.0.0" from @apollo/client@3.4.17
       npm ERR! node_modules/@apollo/client
       npm ERR!   @apollo/client@"3.4.17" from the root project
       npm ERR! 
       npm ERR! Fix the upstream dependency conflict, or retry
       npm ERR! this command with --force, or --legacy-peer-deps
       npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
       npm ERR! 
       npm ERR! See /tmp/npmcache.j6WPn/eresolve-report.txt for a full report.
       
       npm ERR! A complete log of this run can be found in:
       npm ERR!     /tmp/npmcache.j6WPn/_logs/2021-11-12T05_58_02_132Z-debug.log
-----> Build failed
       
       We're sorry this build is failing! You can troubleshoot common issues here:
       https://devcenter.heroku.com/articles/troubleshooting-node-deploys
       
       Some possible problems:
       
       - Dangerous semver range (>) in engines.node
         https://devcenter.heroku.com/articles/nodejs-support#specifying-a-node-js-version
       
       Love,
       Heroku
       
 !     Push rejected, failed to compile React.js (create-react-app) multi app.
 !     Push failed
```

https://github.com/saleor/react-storefront/blob/2f3de82301435e5fe7698063e8cd9a9ec1075278/package.json#L6

So I change to follow saleor-dashboard configuration
```
    "node": ">=14 <15",
    "npm": ">=6.11.0 <7"
```
And the build succeed.

```
-----> Installing binaries
       engines.node (package.json):  >=14 <15
       engines.npm (package.json):   >=6.11.0 <7
       
       Resolving node version >=14 <15...
       Downloading and installing node 14.18.1...
       Bootstrapping npm >=6.11.0 <7 (replacing 6.14.15)...
       npm >=6.11.0 <7 installed
       
-----> Restoring cache
       Caching has been disabled because NODE_MODULES_CACHE=false
       
-----> Installing dependencies
       Installing node modules (package.json)
       
       > sharp@0.29.2 install /tmp/build_758d4938/node_modules/sharp
       > (node install/libvips && node install/dll-copy && prebuild-install) || (node install/can-compile && node-gyp rebuild && node install/dll-copy)
       
       sharp: Downloading https://github.com/lovell/sharp-libvips/releases/download/v8.11.3/libvips-8.11.3-linux-x64.tar.br
       
       > core-js-pure@3.19.1 postinstall /tmp/build_758d4938/node_modules/core-js-pure
       > node -e "try{require('./postinstall')}catch(e){}"
       
       added 1108 packages from 853 contributors and audited 1118 packages in 31.74s
       
       122 packages are looking for funding
         run `npm fund` for details
       
       found 3 moderate severity vulnerabilities
         run `npm audit fix` to fix them, or `npm audit` for details
       
-----> Build
       Detected both "build" and "heroku-postbuild" scripts
       Running heroku-postbuild
       
       > react-storefront@0.1.0 heroku-postbuild /tmp/build_758d4938
       > npm run build
       
       
       > react-storefront@0.1.0 build /tmp/build_758d4938
       > next build
       
       info  - Loaded env from /tmp/build_758d4938/.env
warn  - SWC minify beta enabled. https://nextjs.org/docs/messages/swc-minify-enabled
       warn  - No build cache found. Please configure build caching for faster rebuilds. Read more: https://nextjs.org/docs/messages/no-cache
       Attention: Next.js now collects completely anonymous telemetry regarding usage.
       This information is used to shape Next.js' roadmap and prioritize features.
       You can learn more, including how to opt-out if you'd not like to participate in this anonymous program, by visiting the following URL:
       https://nextjs.org/telemetry
       
       info  - Checking validity of types...
       info  - Creating an optimized production build...
warn - You have enabled the JIT engine which is currently in preview.
warn - Preview features are not covered by semver, may introduce breaking changes, and can change at any time.
       info  - Compiled successfully
       info  - Collecting page data...
       info  - Generating static pages (0/57)
       info  - Generating static pages (14/57)
       info  - Generating static pages (28/57)
Rich text data not in the EditorJS format.
       info  - Generating static pages (42/57)
       info  - Generating static pages (57/57)
       info  - Finalizing page optimization...
       
       Page                                              Size     First Load JS
       ┌ ● /                                             848 B           200 kB
       ├   /_app                                         0 B             133 kB
       ├ λ /[sitemap]                                    244 B           133 kB
       ├ ○ /404                                          2.3 kB          163 kB
       ├ ○ /account/addressBook                          589 B           197 kB
       ├ ○ /account/login                                1.55 kB         143 kB
       ├ ○ /account/orders                               646 B           197 kB
       ├ ● /account/orders/[token]                       1.37 kB         198 kB
       ├ ○ /account/preferences                          2.13 kB         199 kB
       ├ ○ /account/register                             1.54 kB         143 kB
       ├ ○ /cart                                         1.51 kB         201 kB
       ├ ● /category/[slug]                              1.32 kB         201 kB
       ├   ├ /en-US/category/accessories
       ├   ├ /en-US/category/groceries
       ├   ├ /en-US/category/apparel
       ├   └ [+9 more paths]
       ├ ○ /checkout                                     1.21 kB         200 kB
       ├ ● /collection/[slug]                            1.33 kB         201 kB
       ├   ├ /en-US/collection/featured
       ├   ├ /en-US/collection/featured-products
       ├   ├ /en-US/collection/special-products
       ├   └ [+2 more paths]
       ├ ○ /order                                        472 B           197 kB
       ├ ● /page/[slug]                                  512 B           197 kB
       ├ ● /products/[slug] (ISR: 60 Seconds) (3216 ms)  3.7 kB          203 kB
       ├   ├ /en-US/products/saleor-beanie (840 ms)
       ├   ├ /en-US/products/less-is-mo
       ├   ├ /en-US/products/polo-shirt
       ├   ├ /en-US/products/polo-shirt-2
       ├   ├ /en-US/products/saleor-cup
       ├   ├ /en-US/products/saleor-digital-pack
       ├   ├ /en-US/products/saleor-hoodie
       ├   └ [+6 more paths]
       └ ○ /search                                       1.36 kB         198 kB
       + First Load JS shared by all                     133 kB
         ├ chunks/framework-beb51d85c0b60541.js          42.1 kB
         ├ chunks/main-9f9209d678e95caa.js               28.7 kB
         ├ chunks/pages/_app-b1e6a06895e54cab.js         61.3 kB
         ├ chunks/webpack-29f4e7141b68b4d4.js            831 B
         └ css/4221126af8035ac7.css                      7.08 kB
       
       λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
       ○  (Static)  automatically rendered as static HTML (uses no initial props)
       ●  (SSG)     automatically generated as static HTML + JSON (uses getStaticProps)
          (ISR)     incremental static regeneration (uses revalidate in getStaticProps)
       
       
       > react-storefront@0.1.0 postbuild /tmp/build_758d4938
       > next-sitemap
       
       Loaded env from /tmp/build_758d4938/.env
       
-----> Pruning devDependencies
       Skipping because NPM_CONFIG_PRODUCTION is 'false'
       
-----> Build succeeded!
=====> Downloading Buildpack: https://github.com/mars/create-react-app-inner-buildpack.git#v9.0.0
=====> Detected Framework: React.js (create-react-app)
       Writing `static.json` to support create-react-app
       Enabling runtime environment variables
=====> Downloading Buildpack: https://github.com/heroku/heroku-buildpack-static.git
=====> Detected Framework: Static HTML
-----> Installed nginx 1.19.0 to /app/bin
Using release configuration from last framework (Static HTML).
-----> Discovering process types
       Procfile declares types     -> (none)
       Default types for buildpack -> web
-----> Compressing...
       Done: 134.2M
-----> Launching...
```

But when I open the application, it shows nginx 404 not found. And when I check the log, it shows

```
2021-11-12T06:18:51.491389+00:00 app[web.1]: ls: cannot access '/app/build/static/js/*.js': No such file or directory
2021-11-12T06:18:51.491653+00:00 app[web.1]: Error injecting runtime env: bundle not found '/app/build/static/js/*.js'. See: https://github.com/mars/create-react-app-buildpack/blob/master/README.md#user-content-custom-bundle-location
2021-11-12T06:18:51.750685+00:00 app[web.1]: Starting log redirection...
2021-11-12T06:18:51.750932+00:00 app[web.1]: Starting nginx...
```

I try to connect to heroku console and I don't found /app/
![image](https://user-images.githubusercontent.com/476869/141420984-c0a327ba-5cdd-4c68-86f8-4806d6246fb8.png)

Is there something wrong with the configuration?

**Comment by niwasmala** [comment]
Okay, I got it working. Some changes I made.

**Don't use `mars/create-react-app` as buildpacks. Use `heroku/nodejs` and run Next as nodejs apps**

At first, I assume that the storefront is like dashboard, build react static pages and deploy it. I try to build next to static pages by changing the package.json build script from "next build" to "next build && next export" and adding static.json to read from "out/" folder as mentioned in [Next documentation](https://nextjs.org/docs/advanced-features/static-html-export), but build failed with error
```
Error: i18n support is not compatible with next export. See here for more info on deploying: https://nextjs.org/docs/deployment
    at /tmp/build_7bd606c1/node_modules/next/dist/export/index.js:247:19
    at async Span.traceAsyncFn (/tmp/build_7bd606c1/node_modules/next/dist/trace/trace.js:74:20)
```

So I though why not running Next as nodejs application.

This is what I do.
1. Change the buildpack to `heroku/nodejs`
2. Add start script in package json `"start": "next start -p $PORT"`

Hope it helps.

https://docs.saleor.io/docs/2.11/developer/running-saleor/heroku#storefront-and-dashboard-deployment
Perhaps the documentation needs update?
